test/runtests.jl

```diff
@@ -1,82 +1,120 @@
 using DiffRules
 using Test
+using FiniteDifferences
+
+using IrrationalConstants: fourπ
 
 import SpecialFunctions, NaNMath, LogExpFunctions
 import Random
 Random.seed!(1)
 
-function finitediff(f, x)
-    ϵ = cbrt(eps(typeof(x))) * max(one(typeof(x)), abs(x))
-    return (f(x + ϵ) - f(x - ϵ)) / (ϵ + ϵ)
-end
+# Set `max_range` to avoid domain errors.
+const finitediff = central_fdm(5, 1, max_range=1e-3)
 
 @testset "DiffRules" begin
 @testset "check rules" begin
 
 non_diffeable_arg_functions = [(:Base, :rem2pi, 2), (:Base, :ldexp, 2), (:Base, :ifelse, 3)]
 
-for (M, f, arity) in DiffRules.diffrules(; filter_modules=nothing)
-    (M, f, arity) ∈ non_diffeable_arg_functions && continue
-    if arity == 1
-        @test DiffRules.hasdiffrule(M, f, 1)
-        deriv = DiffRules.diffrule(M, f, :goo)
-        modifier = if f in (:asec, :acsc, :asecd, :acscd, :acosh, :acoth)
-            1.0
-        elseif f === :log1mexp
-            -1.0
-        elseif f === :log2mexp
-            -0.5
-        else
-            0.0
-        end
-        @eval begin
-            let
-                goo = rand() + $modifier
-                @test isapprox($deriv, finitediff($M.$f, goo), rtol=0.05)
-                # test for 2pi functions
-                if "mod2pi" == string($M.$f)
-                    goo = 4pi + $modifier
-                    @test NaN === $deriv
+@testset "($M, $f, $arity)" for (M, f, arity) in DiffRules.diffrules(; filter_modules=nothing)
+    for T in [Float32, Float64]
+        (M, f, arity) ∈ non_diffeable_arg_functions && continue
+        if arity == 1
+            @test DiffRules.hasdiffrule(M, f, 1)
+            deriv = DiffRules.diffrule(M, f, :goo)
+            @eval begin
+                let
+                    goo = if $(f in (:asec, :acsc, :asecd, :acscd, :acosh, :acoth))
+                        # avoid singularities with finite differencing
+                        rand($T) + $T(1.5)
+                    elseif $(f in (:log, :airyaix, :airyaiprimex))
+                        # avoid singularities with finite differencing
+                        rand($T) + $T(0.5)
+                    elseif $(f === :log1mexp)
+                        rand($T) - one($T)
+                    elseif $(f in (:log2mexp, :erfinv))
+                        rand($T) - $T(0.5)
+                    else
+                        rand($T)
+                    end
+                    # We're happy with types with the correct promotion behavior, e.g.
+                    # it's fine to return `1` as a derivative despite input being `Float64`.
+                    @test one($T) * $deriv isa $T
```

Comment:
A bit more direct maybe:
```suggestion
                    @test promote_type(typeof($deriv), $T) === $T
```